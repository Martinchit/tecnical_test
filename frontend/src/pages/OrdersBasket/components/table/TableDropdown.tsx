import React from 'react';
import { Dropdown } from 'react-bootstrap';
import { TableDropdownButton } from './TableDropdownButton';
import { OrderExecutionMode } from '../../../../types/types';
import { ExecutionMode } from '../../../../core/lib/utils/executionMode';

interface TableDropdownProps {
  title: OrderExecutionMode | undefined;
  dropdownID: string;
  onSelect: (id: number) => void;
  disabled: boolean;
}

export const TableDropdown: React.FC<TableDropdownProps> = ({
  title,
  dropdownID,
  onSelect,
  disabled,
}) => (
  <Dropdown onSelect={(id: string) => onSelect(Number(id))}>
    <TableDropdownButton
      className="pull-right"
      disabled={disabled}
      variant="light"
      id={dropdownID}
    >
      {title}
    </TableDropdownButton>
    <Dropdown.Menu>
      {ExecutionMode.map((option, i) => (
        <Dropdown.Item eventKey={`${i}`} key={option}>
          {option}
        </Dropdown.Item>
      ))}
    </Dropdown.Menu>
  </Dropdown>
);
