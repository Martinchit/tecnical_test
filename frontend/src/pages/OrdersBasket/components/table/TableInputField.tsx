import styled from 'styled-components';
import { FormControl } from 'react-bootstrap';

export const TableInputField = styled(FormControl)`
  width: 100%;
`;
