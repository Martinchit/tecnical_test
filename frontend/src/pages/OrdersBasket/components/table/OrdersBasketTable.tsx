import React from 'react';
import { Form, InputGroup, Button } from 'react-bootstrap';
import { TableContainer } from './TableContainer';
import { TableBadge } from './TableBadge';
import { TableDropdown } from './TableDropdown';
import { TableInputField } from './TableInputField';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';
import { TableErrorContainer } from '../../../../core/components/atoms/TableErrorContainer';
import { TableCell } from '../../../../core/components/atoms/TableCell';
import { TableHead } from '../../../../core/components/atoms/TableHead';
import { Alert } from '../../../../core/components/atoms/Alert';
import { StockOrder } from '../../../../types/types';

interface OrdersBasketTableProps {
  orders: StockOrder[];
  onSelectCheckClickHandler: (orderId: string, selected: boolean) => void;
  onExecutionModeSelectHandler: (stockIdx: number, modeIdx: number) => void;
  onOrderPriceChangeHandler: (stockIdx: number, orderPrice: string) => void;
  onShareAmountChangeHandler: (stockIdx: number, shareAmount: string) => void;
  onCancelButtonClickHandler: (stockIdx: number) => void;
}

export const OrdersBasketTable: React.FC<OrdersBasketTableProps> = ({
  onSelectCheckClickHandler,
  onExecutionModeSelectHandler,
  onOrderPriceChangeHandler,
  onShareAmountChangeHandler,
  onCancelButtonClickHandler,
  orders,
}) => {
  return (
    <TableContainer responsive>
      <thead>
        <tr>
          <TableHead />
          <TableHead>Status</TableHead>
          <TableHead>Side</TableHead>
          <TableHead>Stock Code</TableHead>
          <TableHead>Execution Code</TableHead>
          <TableHead>Order Price</TableHead>
          <TableHead>Amount &#40;Shares&#41;</TableHead>
        </tr>
      </thead>
      <tbody>
        {orders.map(
          (
            {
              id,
              side,
              bloombergTickerLocal,
              status,
              executionMode,
              currency,
              orderPrice,
              shareAmount,
              error,
              selected,
              orderId,
            },
            i
          ) => {
            const disabledOnMarketMode =
              executionMode !== undefined && executionMode === 'Market';
            const disabledOnSubmittedCondition =
              status !== 'Not Ready' && status !== 'Ready';
            return (
              <React.Fragment key={orderId}>
                <tr>
                  <TableCell>
                    <Form.Check
                      checked={selected}
                      onChange={() => onSelectCheckClickHandler(orderId, selected)}
                    />
                  </TableCell>
                  <TableCell>
                    <TableBadge orderStatus={status} />
                  </TableCell>
                  <TableCell>{side}</TableCell>
                  <TableCell>{bloombergTickerLocal}</TableCell>
                  <TableCell>
                    <TableDropdown
                      dropdownID={`${id}_${side}_mode_dropdown`}
                      title={executionMode}
                      disabled={disabledOnSubmittedCondition}
                      onSelect={(optionId) =>
                        onExecutionModeSelectHandler(i, Number(optionId))
                      }
                    />
                  </TableCell>
                  <TableCell>
                    <InputGroup>
                      <TableInputField
                        disabled={
                          disabledOnMarketMode || disabledOnSubmittedCondition
                        }
                        value={orderPrice || ''}
                        type="number"
                        aria-describedby={`${id}_${side}_input_append`}
                        onChange={({ target }) =>
                          onOrderPriceChangeHandler(i, target.value)
                        }
                      />
                      <InputGroup.Append id={`${id}_${side}_input_append`}>
                        <InputGroup.Text>{currency}</InputGroup.Text>
                      </InputGroup.Append>
                    </InputGroup>
                  </TableCell>
                  <TableCell>
                    <TableInputField
                      disabled={disabledOnSubmittedCondition}
                      value={shareAmount || ''}
                      type="number"
                      onChange={({ target }) =>
                        onShareAmountChangeHandler(i, target.value)
                      }
                    />
                  </TableCell>
                </tr>
                {error ? (
                  <tr>
                    <TableErrorContainer colSpan={7}>
                      <Alert variant="danger">
                        <FlexBox justifyContent="space-between">
                          {`Error: ${error}`}
                          <Button
                            variant="light"
                            onClick={() => onCancelButtonClickHandler(i)}
                          >
                            Close
                          </Button>
                        </FlexBox>
                      </Alert>
                    </TableErrorContainer>
                  </tr>
                ) : null}
              </React.Fragment>
            );
          }
        )}
      </tbody>
    </TableContainer>
  );
};
