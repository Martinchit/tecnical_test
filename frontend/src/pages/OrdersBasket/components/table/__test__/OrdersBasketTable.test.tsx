import React from 'react';
import { shallow } from 'enzyme';
import { Form, Button } from 'react-bootstrap';

import { OrdersBasketTable } from '../OrdersBasketTable';
import { TableDropdown } from '../TableDropdown';
import { TableInputField } from '../TableInputField';
import {
  OrderSide,
  OrderStatus,
  ErrorResponse,
  OrderExecutionMode,
} from '../../../../../types/types';

describe('OrdersBasketTable', () => {
  const orderId = '9cc4ca45-b42b-462d-a8aa-931e6041cfcb';
  const props = {
    onSelectCheckClickHandler: jest.fn(),
    onExecutionModeSelectHandler: jest.fn(),
    onOrderPriceChangeHandler: jest.fn(),
    onShareAmountChangeHandler: jest.fn(),
    onCancelButtonClickHandler: jest.fn(),
    orders: [
      {
        id: 1,
        stockId: 'testID',
        currency: 'HKD',
        bloombergTickerLocal: 'testTicker',
        side: 'Buy' as OrderSide,
        status: 'Not Ready' as OrderStatus,
        selected: false,
        orderId,
      },
    ],
  };

  const restoreMock = () => {
    props.onSelectCheckClickHandler.mockRestore();
    props.onExecutionModeSelectHandler.mockRestore();
    props.onOrderPriceChangeHandler.mockRestore();
    props.onShareAmountChangeHandler.mockRestore();
    props.onCancelButtonClickHandler.mockRestore();
  };

  it('renders correctly', () => {
    const wrapper = shallow(<OrdersBasketTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles onSelectCheckClickHandler', () => {
    const wrapper = shallow(<OrdersBasketTable {...props} />);
    wrapper.find(Form.Check).first().simulate('change');
    expect(props.onSelectCheckClickHandler).toHaveBeenCalled();
    expect(props.onSelectCheckClickHandler).toHaveBeenCalledWith(
      props.orders[0].orderId,
      props.orders[0].selected
    );
    restoreMock();
  });

  it('handles onExecutionModeSelectHandler', () => {
    const optionId = 1;
    const wrapper = shallow(<OrdersBasketTable {...props} />);
    wrapper.find(TableDropdown).first().simulate('select', optionId);
    expect(props.onExecutionModeSelectHandler).toHaveBeenCalled();
    expect(props.onExecutionModeSelectHandler).toHaveBeenCalledWith(0, optionId);
    restoreMock();
  });

  it('handles onOrderPriceChangeHandler', () => {
    const event = { target: { value: 'test' } };
    const wrapper = shallow(<OrdersBasketTable {...props} />);
    wrapper.find(TableInputField).first().simulate('change', event);
    expect(props.onOrderPriceChangeHandler).toHaveBeenCalled();
    expect(props.onOrderPriceChangeHandler).toHaveBeenCalledWith(
      0,
      event.target.value
    );
    restoreMock();
  });

  it('handles onShareAmountChangeHandler', () => {
    const event = { target: { value: 'test' } };
    const wrapper = shallow(<OrdersBasketTable {...props} />);
    wrapper.find(TableInputField).last().simulate('change', event);
    expect(props.onShareAmountChangeHandler).toHaveBeenCalled();
    expect(props.onShareAmountChangeHandler).toHaveBeenCalledWith(
      0,
      event.target.value
    );
    restoreMock();
  });

  it('handles onCancelButtonClickHandler', () => {
    const p = {
      ...props,
      orders: [
        {
          id: 1,
          stockId: 'testID',
          currency: 'HKD',
          bloombergTickerLocal: 'testTicker',
          side: 'Buy' as OrderSide,
          status: 'Not Ready' as OrderStatus,
          error: 'Test Error',
          selected: false,
          orderId,
        },
      ],
    };
    const wrapper = shallow(<OrdersBasketTable {...p} />);
    wrapper.find(Button).last().simulate('click');
    expect(props.onCancelButtonClickHandler).toHaveBeenCalled();
    restoreMock();
  });

  it('disables dropdown and order price input field', () => {
    const p = {
      ...props,
      orders: [
        {
          id: 1,
          stockId: 'testID',
          currency: 'HKD',
          bloombergTickerLocal: 'testTicker',
          side: 'Buy' as OrderSide,
          status: 'Not Ready' as OrderStatus,
          error: 'Test Error',
          executionMode: 'Market' as OrderExecutionMode,
          selected: false,
          orderId,
        },
      ],
    };
    const wrapper = shallow(<OrdersBasketTable {...p} />);
    wrapper.find(Button).last().simulate('click');
    expect(wrapper.find(TableInputField).first().props().disabled).toBeTruthy();
    restoreMock();
  });

  it('disables dropdown and input field when status is not ready nor not ready', () => {
    const p = {
      ...props,
      orders: [
        {
          id: 1,
          stockId: 'testID',
          currency: 'HKD',
          bloombergTickerLocal: 'testTicker',
          side: 'Buy' as OrderSide,
          status: 'Booked' as OrderStatus,
          error: 'Test Error',
          selected: false,
          orderId,
        },
      ],
    };
    const wrapper = shallow(<OrdersBasketTable {...p} />);
    wrapper.find(Button).last().simulate('click');
    expect(wrapper.find(TableDropdown).first().props().disabled).toBeTruthy();
    expect(wrapper.find(TableInputField).first().props().disabled).toBeTruthy();
    expect(wrapper.find(TableInputField).last().props().disabled).toBeTruthy();
    restoreMock();
  });
});
