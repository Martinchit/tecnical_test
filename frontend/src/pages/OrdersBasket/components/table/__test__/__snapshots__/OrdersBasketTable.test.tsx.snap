// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersBasketTable renders correctly 1`] = `
<Styled(Component)
  responsive={true}
>
  <thead>
    <tr>
      <styled.th />
      <styled.th>
        Status
      </styled.th>
      <styled.th>
        Side
      </styled.th>
      <styled.th>
        Stock Code
      </styled.th>
      <styled.th>
        Execution Code
      </styled.th>
      <styled.th>
        Order Price
      </styled.th>
      <styled.th>
        Amount (Shares)
      </styled.th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <styled.td>
        <FormCheck
          checked={false}
          disabled={false}
          inline={false}
          isInvalid={false}
          isValid={false}
          onChange={[Function]}
          title=""
          type="checkbox"
        />
      </styled.td>
      <styled.td>
        <Component
          orderStatus="Not Ready"
        />
      </styled.td>
      <styled.td>
        Buy
      </styled.td>
      <styled.td>
        testTicker
      </styled.td>
      <styled.td>
        <Component
          disabled={false}
          dropdownID="1_Buy_mode_dropdown"
          onSelect={[Function]}
        />
      </styled.td>
      <styled.td>
        <InputGroup>
          <Styled(FormControl)
            aria-describedby="1_Buy_input_append"
            disabled={false}
            onChange={[Function]}
            type="number"
            value=""
          />
          <InputGroupAppend
            id="1_Buy_input_append"
          >
            <InputGroupText>
              HKD
            </InputGroupText>
          </InputGroupAppend>
        </InputGroup>
      </styled.td>
      <styled.td>
        <Styled(FormControl)
          disabled={false}
          onChange={[Function]}
          type="number"
          value=""
        />
      </styled.td>
    </tr>
  </tbody>
</Styled(Component)>
`;
